CDTDatastore/HTTP/CDTRequestLimitInterceptor.m

```diff
@@ -0,0 +1,86 @@
+//
+//  CDTRequestLimitInterceptor.m
+//  CDTDatastore
+//
+//  Created by tomblench on 23/06/2016.
+//  Copyright © 2016 IBM. All rights reserved.
+//
+
+#import "CDTRequestLimitInterceptor.h"
+
+static NSString *kSleepKey = @"com.cloudant.CDTRequestLimitInterceptor.sleep";
+static NSString *kRequestCountKey = @"com.cloudant.CDTRequestLimitInterceptor.requestCount";
+
+@interface CDTRequestLimitInterceptor ()
+
+// the initial time to sleep on receipt of a 429
+@property (readonly) NSTimeInterval initialSleep;
+
+// the maximum number of requests to make on receipt of a 429
+@property (readonly) int maxRequests;
+
+@end
+
+@implementation CDTRequestLimitInterceptor
+
++ (instancetype)interceptor
+{
+    return [[CDTRequestLimitInterceptor alloc] init];
+}
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _initialSleep = 0.25; // 250ms
+        _maxRequests = 3;
+    }
+    return self;
+}
+
+- (instancetype)initWithSleep:(NSTimeInterval)initialSleep
+                  maxRequests:(int)maxRequests;
+{
+    if (self = [super init]) {
+        _initialSleep = initialSleep;
+        _maxRequests = maxRequests;
+    }
+    return self;
+}
+
+/**
+ * Interceptor to retry after an exponential backoff if we receive a 429 error
+ */
+- (CDTHTTPInterceptorContext *)interceptResponseInContext:(CDTHTTPInterceptorContext *)context
+{
+    if (context.response.statusCode == 429) {
+
+        // if we are the first invocation in this pipeline, set some state
+        if (!context.state[kSleepKey]) {
+            [context.state setValue:@(self.initialSleep) forKey:kSleepKey];
+        }
+        if (!context.state[kRequestCountKey]) {
+            [context.state setValue:@0 forKey:kRequestCountKey];
+        }
+        
+        double sleep = [(NSNumber*)context.state[kSleepKey] doubleValue];
+        int requestCount = [(NSNumber*)context.state[kRequestCountKey] intValue];
+
+        if (requestCount >= self.maxRequests) {
```

Comment:
As discussed there is a slight gotcha here in that if the `self.maxRequests` is greater than the initial value of `_remainingRetries` in `CDTURLSessionTask` then this interceptor could log that it will retry, but the retry will never execute because `_remainingRetries` would be exhausted.
This could look a bit confusing in the logs, but I guess the alternative is also checking the number of remaining retries on the session task which means exposing that additional state too.